src/vis/density/HexagonalBin.tsx

```diff
@@ -0,0 +1,411 @@
+  }, [currentColorColumn, currentX, d3Hexbin, xScale, yScale, currentY]);
+
+  // simple radius scale for the hexes
+  const radiusScale = useMemo(() => {
+    if (colsStatus === 'success') {
+      const min = d3v7.min(hexes.map((h) => h.length));
+      const max = d3v7.max(hexes.map((h) => h.length));
+
+      return d3v7
+        .scaleLinear()
+        .domain([min, max])
+        .range([config.hexRadius / 2, config.hexRadius]);
+    }
+
+    return null;
+  }, [colsStatus, hexes, config.hexRadius]);
+
+  // simple opacity scale for the hexes
+  const opacityScale = useMemo(() => {
+    if (colsStatus === 'success') {
+      const min = d3v7.min(hexes.map((h) => h.length));
+      const max = d3v7.max(hexes.map((h) => h.length));
+
+      return d3v7.scaleLinear().domain([min, max]).range([0.1, 1]);
+    }
+
+    return null;
+  }, [colsStatus, hexes]);
+
+  // Create a default color scale
+  const colorScale = useMemo(() => {
+    if (colsStatus !== 'success' || !currentColorColumn?.allValues) {
+      return null;
+    }
+
+    const colorOptions = currentColorColumn.allValues.map((val) => val.val as string);
+
+    return d3v7.scaleOrdinal<string, string>(d3v7.schemeCategory10).domain(Array.from(new Set<string>(colorOptions)));
+  }, [colsStatus, currentColorColumn?.allValues]);
+
+  // memoize the actual hexes since they do not need to change on zoom/drag
+  const hexObjects = React.useMemo(() => {
+    return (
+      <>
+        {hexes.map((singleHex) => {
+          return (
+            <SingleHex
+              key={`${singleHex.x}, ${singleHex.y}`}
+              selected={selected}
+              hexbinOption={config.hexbinOptions}
+              hexData={singleHex}
+              d3Hexbin={d3Hexbin}
+              isSizeScale={config.isSizeScale}
+              radiusScale={radiusScale}
+              isOpacityScale={config.isOpacityScale}
+              opacityScale={opacityScale}
+              hexRadius={config.hexRadius}
+              colorScale={colorScale}
+            />
+          );
+        })}
+      </>
+    );
+  }, [colorScale, config.hexRadius, config.isOpacityScale, config.isSizeScale, d3Hexbin, hexes, opacityScale, radiusScale, selected, config.hexbinOptions]);
+
+  // // apply zoom/panning
+  useEffect(() => {
+    const zoom = d3v7.zoom();
+
+    if (!xScale || !yScale || config.dragMode === EScatterSelectSettings.RECTANGLE) {
+      return;
+    }
+
+    zoom.on('zoom', (event: D3ZoomEvent<any, any>) => {
+      const { transform } = event;
+
+      const newX = transform.rescaleX(xScale);
+      const newY = transform.rescaleY(yScale);
+
+      // Question: I dont think this should be a ref, because it doesnt actually cause a re render. Only the other setters below make it work, if i moved them above this code there would be bugs.
+      // But when I made it a useState object it didnt work with the object.
+      xZoomedScale.current = newX;
+      yZoomedScale.current = newY;
```

Comment:
It's hard for me to figure this out from the code. I think a direct conversation about this point would make sense.